### PR TITLE
#30 articles

### PR DESCRIPTION
**Pull request by Banovvv**
Article related models, repositories, services and two additional controllers

### CODE REVIEW

File: SportsHub/SportsHub.Api/Controllers/ArticleController.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    [Route("api/[controller]")]
+    [ApiController]
+    public class ArticleController : ControllerBase
+    {
+        private readonly IArticleService service;
+        private readonly IMapper mapper;
+
+        public ArticleController(IArticleService service, IMapper mapper)
+        {
+            this.service = service;
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.Api/Controllers/ArticleController.cs:18]: Please do not use `this`. The recommended name for private readonly fields should start from `_`, e.g. _service.
Also, it is not clear from this name what service. It is better to call it `_articleService`.

File: SportsHub/SportsHub.Api/Controllers/ArticleController.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            var article = await service.GetByTitleAsync(title);
+
+            if (article == null)
+            {
+                return BadRequest($"No such article");
+            }
+
+            ArticleResponseDTO articleResponse = mapper.Map<ArticleResponseDTO>(article);
+            return Ok($"{JsonSerializer.Serialize(articleResponse)}");
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.Api/Controllers/ArticleController.cs:33]: Did you try not to use a `JsonSerializer` here?

File: SportsHub/SportsHub.Api/Program.cs [hunk 3]
```diff
@@ -81,13 +81,16 @@
...
 builder.Services.AddTransient<IUserService, UserService>();
+builder.Services.AddTransient<IArticleService, ArticleService>();
+builder.Services.AddTransient<ICommentService, CommentService>();
 builder.Services.AddTransient<IAuthenticationService, AuthenticationService>();
 builder.Services.AddTransient<IJsonTokenService, JsonTokenService>();
 builder.Services.AddTransient<ExceptionHandler>();
 
 //Adding AutoMapper
 //Looks in the assembly the file is located for mapping profiles.
 builder.Services.AddAutoMapper(typeof(UserMapping));
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.Api/Program.cs:92]: There is an overload of this method which gets Assembly collection. Please take a look at it.

File: SportsHub/SportsHub.DAL/Data/ApplicationDbContext.cs [hunk 4]
```diff
@@ -25,6 +29,10 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
         {
             new UserEntityTypeConfiguration().Configure(modelBuilder.Entity<User>());
```
Comment (MBocharskyi) [inline SportsHub/SportsHub.DAL/Data/ApplicationDbContext.cs:30]: Please take a look at `ApplyConfigurationsFromAssembly` method in `ModelBuilder` class.

File: SportsHub/SportsHub.Api/Controllers/ArticleController.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    [ApiController]
+    public class ArticleController : ControllerBase
+    {
+        private readonly IArticleService service;
+        private readonly IMapper mapper;
+
+        public ArticleController(IArticleService service, IMapper mapper)
+        {
+            this.service = service;
+            this.mapper = mapper;
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Controllers/ArticleController.cs:19]: rename private fields, start should be with _, and do it everywhere! remove "this". I think I with Mykola told about it, already.

File: SportsHub/SportsHub.Api/Controllers/ArticleController.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        [HttpGet("GetArticleByTitle")]
+        public async Task<IActionResult> GetArticleByTitleAsync(string title)
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Controllers/ArticleController.cs:23]: I am not sure that we need GetArtickleByTitle here, it looks like too much.

Comment (NataliiaBidiuk) [reply SportsHub/SportsHub.Api/Controllers/ArticleController.cs:23]: @MBocharskyi What do you think about not returning an ActionResult but some response model? Or we will improve it with the middleware of exception?

File: SportsHub/SportsHub.Api/Controllers/ArticleController.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            var article = await service.GetByTitleAsync(title);
+
+            if (article == null)
+            {
+                return BadRequest($"No such article");
+            }
+
+            ArticleResponseDTO articleResponse = mapper.Map<ArticleResponseDTO>(article);
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Controllers/ArticleController.cs:32]: you can, also use 'var' here

File: SportsHub/SportsHub.Application/Authentication/Models/DTOs/ArticleResponseDTO.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    public class ArticleResponseDTO
+    {
+        public int Id { get; set; }
+        public string Title { get; set; }
+        public string Content { get; set; }
+
+        public string StateName { get; set; }
+        public string CategoryName { get; set; }
+
+        public virtual ICollection<User> Authors { get; set; }
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Application/Authentication/Models/DTOs/ArticleResponseDTO.cs:14]: I don't think that you need virtual in Response model

File: SportsHub/SportsHub.Core/Models/Article.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
+﻿namespace SportsHub.Domain.Models
+{
+    public class Article
+    {
+        public Article()
+        {
+            Comments = new HashSet<Comment>();
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Core/Models/Article.cs:7]: why HashSet is here?

File: SportsHub/SportsHub.Application/Services/ArticleService.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        }
+
+        public async Task<Article?> GetByTitleAsync(string title)
+        {
+            return await unitOfWork.ArticleRepository.GetByTitleAsync(title);
+        }
+
+        public async Task<IEnumerable<Article>> GetByAuthorAsync(string author)
+        {
+            return await unitOfWork.ArticleRepository.GetByAuthorAsync(author);
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Application/Services/ArticleService.cs:22]: do we use all these methods?

File: SportsHub/SportsHub.Application/Services/IArticleService.cs [hunk 11]
```diff
@@ -0,0 +1,12 @@
+﻿using SportsHub.Domain.Models;
+
+namespace SportsHub.AppService.Services
+{
+    public interface IArticleService
+    {
+        Task<Article?> GetByTitleAsync(string title);
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Application/Services/IArticleService.cs:7]: is the Title unique? I don`t see that configuration in the DB

File: SportsHub/SportsHub.DAL/Data/Configurations/ArticleEntityTypeConfiguration.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace SportsHub.DAL.Data.Configurations
+{
+    public class ArticleEntityTypeConfiguration : IEntityTypeConfiguration<Article>
+    {
+        public void Configure(EntityTypeBuilder<Article> article)
+        {
+            article.Property(x => x.Title)
+                .IsRequired(true)
+                .HasMaxLength(100)
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.DAL/Data/Configurations/ArticleEntityTypeConfiguration.cs:13]: please, always use Constants for the constraints

File: SportsHub/SportsHub.DAL/Data/Configurations/ArticleEntityTypeConfiguration.cs [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+                .IsUnicode(true);
+
+            article.Property(x => x.Content)
+                .IsRequired(true)
+                .IsUnicode(true);
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.DAL/Data/Configurations/ArticleEntityTypeConfiguration.cs:18]: why in some places IsUnicode set as true and in others as False?

File: SportsHub/SportsHub.DAL/Data/Configurations/CategoryEntityTypeConfiguration.cs [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace SportsHub.DAL.Data.Configurations
+{
+    public class CategoryEntityTypeConfiguration : IEntityTypeConfiguration<Category>
+    {
+        public void Configure(EntityTypeBuilder<Category> category)
+        {
+            category.Property(x => x.Name)
+                .IsRequired(true)
+                .HasMaxLength(50)
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.DAL/Data/Configurations/CategoryEntityTypeConfiguration.cs:13]: const

Comment (NataliiaBidiuk) [reply SportsHub/SportsHub.DAL/Data/Configurations/CategoryEntityTypeConfiguration.cs:13]: do it everywhere

File: SportsHub/SportsHub.DAL/Data/Configurations/CommentEntityTypeConfiguration.cs [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+            comment.Property(x => x.PostedOn)
+                .IsRequired(true)
+                .HasDefaultValueSql("GETDATE()");
+
+            comment.Property(x => x.AuthorId)
+                .IsRequired(true);
+
+            comment.Property(x => x.ArticleId)
+                .IsRequired(true);
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.DAL/Data/Configurations/CommentEntityTypeConfiguration.cs:24]: Please, add navigationProperty and foreign key setup in the configuration as well. I know that it is not necessary, but better to have all configs in the one place. Add it to every dependent tables

File: SportsHub/SportsHub.DAL/Migrations/20221019101502_ArticleAndRelatedEntities.cs [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+{
+    public partial class ArticleAndRelatedEntities : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.AddColumn<int>(
+                name: "ArticleId",
+                table: "Users",
+                type: "int",
+                nullable: true);
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.DAL/Migrations/20221019101502_ArticleAndRelatedEntities.cs:16]: does it a correct migration? I am not sure that we need ArticleId in the User table
Article should keeps an UserId

Comment (NataliiaBidiuk) [reply SportsHub/SportsHub.DAL/Migrations/20221019101502_ArticleAndRelatedEntities.cs:16]: Also, please, don`t do a few migrations for the one commit. It is really hard to review. Drop all added here and then add one new, please.
By the way, better not to have a Pull request with a lot of functionality. It is really hard to review, and there may be missed some important thinks. If you have a lot of issues dependent on each other, just create branches from another branch, not a master!!!

File: SportsHub/SportsHub.DAL/Repository/ArticleRepository.cs [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        public ArticleRepository(ApplicationDbContext context) : base(context)
+        {
+        }
+
+        public async Task<Article?> GetByTitleAsync(string title)
+        {
+            return await DbSet
+                .Where(x => x.Title == title)
+                .FirstOrDefaultAsync();
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.DAL/Repository/ArticleRepository.cs:18]: why first or default? there is a constraint that the Title is unique?

File: SportsHub/SportsHub.DAL/Repository/ArticleRepository.cs [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        public async Task<IEnumerable<Article>> GetByAuthorAsync(string author)
+        {
+            return await DbSet
+                .Where(x => x.Authors.Any(a => a.DisplayName == author))
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.DAL/Repository/ArticleRepository.cs:24]: Let's discuss this feature later. Is it requirement for getting an article by DisplayName?

**Review by NataliiaBidiuk — CHANGES_REQUESTED** [review]

File: SportsHub/SportsHub.Api/Controllers/ArticleController.cs (hunk 5 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Api/Controllers/ArticleController.cs:19]: Fixed

File: SportsHub/SportsHub.DAL/Data/ApplicationDbContext.cs (hunk 4 above)
Comment (Banovvv) [reply SportsHub/SportsHub.DAL/Data/ApplicationDbContext.cs:30]: Implemented `ApplyConfigurationsFromAssembly`

File: SportsHub/SportsHub.Api/Controllers/ArticleController.cs (hunk 7 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Api/Controllers/ArticleController.cs:32]: Fixed

File: SportsHub/SportsHub.Application/Authentication/Models/DTOs/ArticleResponseDTO.cs (hunk 8 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Application/Authentication/Models/DTOs/ArticleResponseDTO.cs:14]: Fixed

File: SportsHub/SportsHub.Core/Models/Article.cs (hunk 9 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Core/Models/Article.cs:7]: To make sure we do not get the same comment more than once

File: SportsHub/SportsHub.Application/Services/ArticleService.cs (hunk 10 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Application/Services/ArticleService.cs:22]: There were no requirements when it came to methods - I wrote several so we have something to work with

File: SportsHub/SportsHub.Application/Services/IArticleService.cs (hunk 11 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Application/Services/IArticleService.cs:7]: Article is unique now.

File: SportsHub/SportsHub.DAL/Data/Configurations/CategoryEntityTypeConfiguration.cs (hunk 14 above)
Comment (Banovvv) [reply SportsHub/SportsHub.DAL/Data/Configurations/CategoryEntityTypeConfiguration.cs:13]: Fixed

File: SportsHub/SportsHub.DAL/Data/Configurations/ArticleEntityTypeConfiguration.cs (hunk 13 above)
Comment (Banovvv) [reply SportsHub/SportsHub.DAL/Data/Configurations/ArticleEntityTypeConfiguration.cs:18]: Unified those.

File: SportsHub/SportsHub.DAL/Data/Configurations/ArticleEntityTypeConfiguration.cs (hunk 12 above)
Comment (Banovvv) [reply SportsHub/SportsHub.DAL/Data/Configurations/ArticleEntityTypeConfiguration.cs:13]: Implemented

File: SportsHub/SportsHub.DAL/Repository/ArticleRepository.cs (hunk 17 above)
Comment (Banovvv) [reply SportsHub/SportsHub.DAL/Repository/ArticleRepository.cs:18]: But `FirstAsync()` will throw and exception if there is no such article.

File: SportsHub/SportsHub.DAL/Data/Configurations/CommentEntityTypeConfiguration.cs (hunk 15 above)
Comment (Banovvv) [reply SportsHub/SportsHub.DAL/Data/Configurations/CommentEntityTypeConfiguration.cs:24]: Done

File: SportsHub/SportsHub.DAL/Migrations/20221019101502_ArticleAndRelatedEntities.cs (hunk 16 above)
Comment (Banovvv) [reply SportsHub/SportsHub.DAL/Migrations/20221019101502_ArticleAndRelatedEntities.cs:16]: Understood

File: SportsHub/SportsHub.Application/Services/ArticleService.cs (hunk 10 above)
Comment (NataliiaBidiuk) [reply SportsHub/SportsHub.Application/Services/ArticleService.cs:22]: I am not sure that it is a good idea, code will hard to support. It is good that you have done more work, but move it to the separate Branch or stash it, please

File: SportsHub/SportsHub.DAL/Repository/ArticleRepository.cs (hunk 17 above)
Comment (NataliiaBidiuk) [reply SportsHub/SportsHub.DAL/Repository/ArticleRepository.cs:18]: I understand, but my question about others: what is we have a few articles with the same title? how do you know that you need the first one?

File: SportsHub/SportsHub.Api/Controllers/ArticleController.cs (hunk 6 above)
Comment (MBocharskyi) [reply SportsHub/SportsHub.Api/Controllers/ArticleController.cs:22]: It is better to use `ActionResult<ResponseModel>`

Comment (Banovvv) [reply SportsHub/SportsHub.Api/Controllers/ArticleController.cs:22]: Like this `public async Task<ActionResult<ArticleResponseDTO>> GetArticleByTitleAsync(string title)`?

File: SportsHub/SportsHub.Application/Services/ArticleService.cs (hunk 10 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Application/Services/ArticleService.cs:22]: Should I stash/branch from everywhere - interfaces, repository, service?

File: SportsHub/SportsHub.Api/Program.cs (hunk 3 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Api/Program.cs:92]: From what I understand, I should have all service interfaces + `IUnitOfWork` implement the same base interface in order to do that, is this correct?

File: SportsHub/SportsHub.Application/Services/ArticleService.cs (hunk 10 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Application/Services/ArticleService.cs:22]: Everything that is not used is now moved to a local branch.

File: SportsHub/SportsHub.Api/Program.cs (hunk 3 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Api/Program.cs:101]: I understood the question was about the services, but it is for AutoMapper. Implemented using this [article](https://dev.to/sumitkharche/how-to-integrate-automapper-in-asp-net-core-web-api-3cn4#:~:text=The%20AddAutoMapper%20method%20provided%20by%20the%20AutoMapper%20package,single%20assembly%20or%20list%20of%20assemblies%20as%20input.).

File: SportsHub/SportsHub.Core/Models/Article.cs (hunk 9 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Core/Models/Article.cs:7]: It is the same [here](https://github.com/mehalyna/Sf-069.Net-Project/blob/17fe27ce3020892f587503a964de1e0414d2c04e/SportsHub/SportsHub.Core/Models/Role.cs#L7) and the pull request was approved. This is the way I know collections are initialized in EF.

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+using Microsoft.AspNetCore.Mvc;
+using SportsHub.AppService.Services;
+
+namespace SportsHub.Api.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class CommentController : ControllerBase
+    {
+        private readonly ICommentService service;
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Controllers/CommentController.cs:11]: _service, _mapper, rename everwhere

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+        private readonly IMapper mapper;
+
+        public CommentController(ICommentService service, IMapper mapper)
+        {
+            this.service = service;
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Api/Controllers/CommentController.cs:16]: remove this, everywhere!

File: SportsHub/SportsHub.Application/Authentication/Models/DTOs/ArticleResponseDTO.cs [hunk 21]
```diff
@@ -0,0 +1,16 @@
+﻿using SportsHub.Domain.Models;
+
+namespace SportsHub.AppService.Authentication.Models.DTOs
+{
+    public class ArticleResponseDTO
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Application/Authentication/Models/DTOs/ArticleResponseDTO.cs:5]: why is it here? response model should be in API layers

File: SportsHub/SportsHub.Application/Services/ArticleService.cs [hunk 22]
```diff
@@ -0,0 +1,20 @@
+﻿using SportsHub.Domain.Models;
+using SportsHub.Domain.UOW;
+
+namespace SportsHub.AppService.Services
+{
+    public class ArticleService : IArticleService
+    {
+        private readonly IUnitOfWork unitOfWork;
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Application/Services/ArticleService.cs:8]: _unitOfWork, remove this.

File: SportsHub/SportsHub.Core/Models/Article.cs (hunk 9 above)
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.Core/Models/Article.cs:7]: I don`t think that we need add empty HashSet here

File: SportsHub/SportsHub.DAL/Data/ApplicationDbContext.cs [hunk 23]
```diff
@@ -16,15 +16,19 @@ public ApplicationDbContext(DbContextOptions options) : base(options)
...
         protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
         {
         }
 
         protected override void OnModelCreating(ModelBuilder modelBuilder)
         {
-            new UserEntityTypeConfiguration().Configure(modelBuilder.Entity<User>());
-            new RoleEntityTypeConfiguration().Configure(modelBuilder.Entity<Role>());
+            modelBuilder.ApplyConfigurationsFromAssembly(Assembly.GetExecutingAssembly(),
+                x => x.GetInterfaces().Any(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IEntityTypeConfiguration<>)));
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.DAL/Data/ApplicationDbContext.cs:31]: this should be enough modelBuilder.ApplyConfigurationsFromAssembly(GetType().Assembly);

File: SportsHub/SportsHub.DAL/Data/Configurations/ArticleEntityTypeConfiguration.cs [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+using SportsHub.DAL.Data.Configurations.Constants;
+using SportsHub.Domain.Models;
+
+namespace SportsHub.DAL.Data.Configurations
+{
+    public class ArticleEntityTypeConfiguration : IEntityTypeConfiguration<Article>
+    {
+        public void Configure(EntityTypeBuilder<Article> article)
+        {
+            article
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.DAL/Data/Configurations/ArticleEntityTypeConfiguration.cs:12]: The title should have a constraint on the length. For all string types better to have limitations for the database!

File: SportsHub/SportsHub.DAL/Data/Configurations/Constants/ConfigurationConstants.cs [hunk 25]
```diff
@@ -0,0 +1,18 @@
+﻿namespace SportsHub.DAL.Data.Configurations.Constants
+{
+    public static class ConfigurationConstants
+    {
+        public const int articleTitleMaxLength = 100;
```
Comment (NataliiaBidiuk) [inline SportsHub/SportsHub.DAL/Data/Configurations/Constants/ConfigurationConstants.cs:5]: Use Uppercase for the first later

File: SportsHub/SportsHub.DAL/Data/ApplicationDbContext.cs (hunk 23 above)
Comment (Banovvv) [reply SportsHub/SportsHub.DAL/Data/ApplicationDbContext.cs:31]: Implemented.

File: SportsHub/SportsHub.DAL/Data/Configurations/Constants/ConfigurationConstants.cs (hunk 25 above)
Comment (Banovvv) [reply SportsHub/SportsHub.DAL/Data/Configurations/Constants/ConfigurationConstants.cs:5]: Fixed all.

File: SportsHub/SportsHub.Core/Models/Article.cs (hunk 9 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Core/Models/Article.cs:7]: Okay, I've removed all `HashSets` from the constructors but everyone should remember to initialize the collections before trying to use them.

File: SportsHub/SportsHub.Application/Services/ArticleService.cs (hunk 22 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Application/Services/ArticleService.cs:8]: fixed

File: SportsHub/SportsHub.Application/Authentication/Models/DTOs/ArticleResponseDTO.cs (hunk 21 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Application/Authentication/Models/DTOs/ArticleResponseDTO.cs:5]: Where in API layer?

File: SportsHub/SportsHub.Api/Controllers/CommentController.cs (hunk 20 above)
Comment (Banovvv) [reply SportsHub/SportsHub.Api/Controllers/CommentController.cs:16]: fixed

**Review by NataliiaBidiuk — APPROVED** [review]

**Pull request merged**